### PR TITLE
V1.12.2 branch

### PR DESCRIPTION
**Pull request by richbolen**
This pull request is a fork of v1.12.2 with added support for SELECT JSON/INSERT JSON and case sensitive table and column names.   You've already implemented INSERT JSON.  I did it slightly differently but it should not be too difficult to port.

### CODE REVIEW

**Comment by alexflav23** [comment]
Hi @richbolen,

Thank you for your contribution, but there are a couple of things off with the pull request:

- You are removing our license which explicitly prevents binary redistributions, unless you are extremely keen that we sit together in court this is perhaps not the right move.
- You are using a very specific JSON library, so it will not be possible to merge directly, but we can use the idea to implement the functionality in a library agnostic way, since there is no way for us to know what kind of JSON library the end user will have.
- I don't really understand what the case sensitive handling stuff is doing, maybe you can explain what the problems you are addressing are.
- Internal logging is all over the place for something that only concerns the framework developers, so we will need to clean that up a bit before merging.
- You have also disabled the bintray plugin and other such changes which probably made sense at the time of the implementation but cannot be merged in to the project without breaking the existing functionality inside phantom.

**Comment by richbolen** [comment]
Sorry again about the license.  That was unintentional.  For the files I created I didn't realize I needed to add it.  

I agree about the JSON lib.  I think the datastax driver just works with strings for the json values, so phantom could do the same.  

Cassandra supports case sensitive keyspace, table, and field names.  But in order to use them, you have to double-quote the names in the CQL calls.  So the code is mainly trying to insure everything is quoted properly in the CQL.  We use them because we want our model objects to exactly match the CQL table structure.  It eases the process of ser/deser to json for inserting and querying.  

Agree again on the logging.  That was me trying to figure out how the driver worked and to insure the case-sensitive stuff was working.  

The bintray thing was just to get it to compile on my local system.  For some reason it did not like the plugin but I have not had time to research it.

Apologies again if this isn't very useful for you.  I am working on porting this to your latest dev branch.  I could port, make the suggested changes, and do another pull request if that will help.

File: project/Build.scala [hunk 1]
```diff
@@ -134,7 +134,7 @@ object Build extends Build {
     testOptions in Test := Seq(Tests.Filter(x => !performanceFilter(x))),
     testOptions in PerformanceTest := Seq(Tests.Filter(x => performanceFilter(x))),
     fork in PerformanceTest := false
-  ) ++ net.virtualvoid.sbt.graph.Plugin.graphSettings ++ publishSettings ++ VersionManagement.newSettings
+  ) ++ VersionManagement.newSettings // ++ publishSettings
```
Comment (alexflav23) [inline project/Build.scala]: Codacy Issue found: [Whitespace at end of line](https://www.codacy.com/app/flavian/phantom_2/file/2347799639/issues/source?bid=2847375&fileBranchId=3216756#l137)

File: phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/JsonUtils.scala [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        Manager.logger.debug(s"Default case => ${default}")
+        if (default != null) Manager.logger.debug(s"with type ${default.getClass().getCanonicalName()}")
+        default
+      }
+    }
+  }
+}
+
+case object JodaDateTimeSerializer extends CustomSerializer[org.joda.time.DateTime](format => ( {
+  case JNull => null
```
Comment (alexflav23) [inline phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/JsonUtils.scala]: Codacy Issue found: [Avoid using null](https://www.codacy.com/app/flavian/phantom_2/file/2347799465/issues/source?bid=2847375&fileBranchId=3216756#l265)

File: phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/JsonUtils.scala [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        case ja: JArray => (unwrap(x._1), unwrapJArray(ja))
+        case js: JString => (unwrap(x._1), js)
+        case ji: JInt => (unwrap(x._1), ji)
+        case jd: JDouble => (unwrap(x._1), jd)
+        case jb: JBool => (unwrap(x._1), jb)
+        case m: Map[String, JValue] => (unwrap(x._1), unwrapJObject(m))
+        case l: List[Any] => (unwrap(x._1), unwrapJArray(l))
+        case default => {
+          Manager.logger.debug(s"Unwrapping default value from JObject ${x._1} => ${default}")
+          if (default != null) Manager.logger.debug(s"with type ${default.getClass().getCanonicalName()}")
```
Comment (alexflav23) [inline phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/JsonUtils.scala]: Codacy Issue found: [Avoid using null](https://www.codacy.com/app/flavian/phantom_2/file/2347799465/issues/source?bid=2847375&fileBranchId=3216756#l94)

File: project/Build.scala (hunk 1 above)
Comment (alexflav23) [inline project/Build.scala]: Codacy Issue found: [Whitespace at end of line](https://www.codacy.com/app/flavian/phantom_2/file/2347799639/issues/source?bid=2847375&fileBranchId=3216756#l137)

File: phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/JsonUtils.scala (hunk 2 above)
Comment (alexflav23) [inline phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/JsonUtils.scala]: Codacy Issue found: [Avoid using null](https://www.codacy.com/app/flavian/phantom_2/file/2347799465/issues/source?bid=2847375&fileBranchId=3216756#l265)

File: phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/JsonUtils.scala (hunk 3 above)
Comment (alexflav23) [inline phantom-dsl/src/main/scala/com/websudos/phantom/builder/query/JsonUtils.scala]: Codacy Issue found: [Avoid using null](https://www.codacy.com/app/flavian/phantom_2/file/2347799465/issues/source?bid=2847375&fileBranchId=3216756#l94)

**Comment by alexflav23** [comment]
Hi @richbolen,

Just to keep you in the loop, all issues have now been internally resolved, we have switched the entire project to use an implicit query builder that is very easily modifiable, some last few issues and tests to update around that but other than that:

- Select json
- Insert json
- Case sensitive table names
- Ability to customise the query builder natively supported.
- USING clause support for ignore nulls

This should all be available in the `1.27.0` release coming up probably mid next week.

Regards.

**Comment by alexflav23** [comment]
Hi @richbolen,

A long overdue update on this.

- [x] Select json
- [x] Insert json
- [x] Ability to customise the query builder natively supported.
- [ ] Case sensitive table names
- [ ] USING clause support for ignore nulls

The fundamental mechanism we used to enable auto-generation of schema and make phantom work has now changed completely in 2.0.0, so most of the tricks in pre 2.0.0 have gone away, which makes this PR and previous work not so useful.

Table names are now derived and injected at compile time using macro compat and implicit macros, with a recursive type search. In English, that means we could have far simpler ways of propagating configuration that don't require complex passing through of builder settings.

In post 2.0.0, this will be available as a very simple config option to somehow be provided compile time, so you won't be able to change this at runtime but I guess there wouldn't really be a lot of demand for doing that anyway.

Regards,
Flavian

**Comment by richbolen** [comment]
Thanks Flavian!  This is excellent.


Rich

________________________________

**Comment by alexflav23** [comment]
Hi @richbolen,

Another overdue update on this, alongside a happy new year!

2.0.0 is now live and available via GitHub/Bintray/Maven Central and the officially supported version of phantom. The last of the mohicans, which is configuration support for controlling how table names are generated, will be coming up in Phantom 2.0.0's next release.

This will be an implicit object that will be looked up by the macro generated code to dictate what will happen compile time, so it will be as single as providing an implicit configuration in the scope of the table, or better yet, in the scope of the `Database` which holds your tables, making it pretty trivial to do all the things you wanted and more.

We have also added supported for the final missing bit, `USING IGNORE NULLS`, which should basically make this whole PR available by default in phantom using all the new cool toys, e.g the macro based inference mechanism that completely replaces the old engine.

Regards,
Flavian

**Comment by alexflav23** [comment]
Hi @richbolen,

Case sensitive configuration is going to make its way to the next release of phantom, which probably means this PR can be closed. Inserting and selecting JSON as well as hopefully everything else requested in here should now be available by default, especially in 2.0.x series of phantom.

Regards.

**Pull request closed without merging**